### PR TITLE
[SPARK-16868][Web Ui] Fix executor be both dead and alive on executor ui.

### PR DESCRIPTION
**Pull request by SaintBacchus**
## What changes were proposed in this pull request?

In a heavy pressure of the spark application, since the executor will register it to driver block manager twice(because of heart beats), the executor will show as picture show:
image
## How was this patch tested?

NA

Details in: [SPARK-16868](https://issues.apache.org/jira/browse/SPARK-16868)

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/StorageStatusListener.scala [hunk 1]
```diff
@@ -77,6 +77,18 @@ class StorageStatusListener(conf: SparkConf) extends SparkListener {
...
       val storageStatus = new StorageStatus(blockManagerId, maxMem)
       executorIdToStorageStatus(executorId) = storageStatus
+
+      // Try to remove the dead storage status if same executor register the block manger twice.
+      removeDeadExecutorStorageStatus(executorId)
+    }
+  }
+
+  private def removeDeadExecutorStorageStatus(executorId: String): Unit = {
+    deadExecutorStorageStatus.zipWithIndex.foreach { case (status, index) =>
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/storage/StorageStatusListener.scala]: Just commenting on the code, not the logic: isn't this just something like `deadExecutorStorageStatus.retain(_.blockManagerId.executorId != executorId)`? I know it does something slightly different, but maybe it's desirable to clear all such matching statuses. And it's a one-liner that doesn't need a method.

Comment (SaintBacchus) [inline core/src/main/scala/org/apache/spark/storage/StorageStatusListener.scala]: `retain` seems to be a method in `MapLike`, but I can't find any similar method in `ListBuffer`.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/storage/StorageStatusListener.scala]: Ah. OK what about 

```
val toRemove = deadExecutorStorageStatus.find(_.blockManagerId.executorId == executorId)
if (toRemove >= 0) {
  deadExecutorStorageStatus.remove(toRemove)
}
```
Might be clearer and more direct. No big deal either way.

**Comment by SaintBacchus** [comment]
\cc  @srowen

**Comment by srowen** [comment]
As to the actual logic of the change, I don't feel that confident about reviewing it. Suggest looking at 'blame' output here to see who might have written it?

**Comment by vanzin** [comment]
Looks fine (small style nit but not worth another review round). I'm a little curious about why it happens - I thought that the driver would kill executors that failed a heartbeat, but that's a separate discussion.

Merging to master.

**Comment by SaintBacchus** [comment]
I will re-run this case, and dig into why the executor will double register.

**Pull request closed without merging**